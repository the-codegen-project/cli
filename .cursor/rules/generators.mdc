---
description: Generator implementation patterns and requirements for The Codegen Project CLI
globs: ["src/codegen/generators/**/*.ts"]
alwaysApply: true
---

# Generator Implementation Rules

## Development Philosophy: Expected Output First

**MANDATORY APPROACH**: Before implementing any generator, you MUST:

1. **Create Expected Output Manually** in `test/runtime/typescript/src/`
2. **Write Runtime Tests** that validate the expected output works
3. **Validate Manual Implementation** passes all tests
4. **Build Generator** to produce identical output
5. **Verify Generated Code** passes the same tests

### Runtime Project Structure
```
test/runtime/typescript/
├── src/                     # Where you manually create expected output
├── test/                    # Where you write tests for expected output
│   ├── payloads.spec.ts     # Tests for payload models
│   ├── parameters.spec.ts   # Tests for parameter models
│   ├── headers.spec.ts      # Tests for header models
│   ├── types.spec.ts        # Tests for type definitions
│   ├── channels/            # Tests for channel functions
│   └── client/              # Tests for client code
├── codegen-*.mjs            # Generator configurations
└── package.json             # Dependencies for generated code
```

### Expected Output Validation Process
```bash
# 1. Create manual implementation
cd test/runtime/typescript/src/
# Create your expected files manually

# 2. Write/adapt tests
cd ../test/
# Create/modify test files

# 3. Validate manual implementation
npm test -- --testNamePattern="your-feature"

# 4. Build generator (only after manual validation)
# Implement generator in src/codegen/generators/

# 5. Test generator output
npm run generate  # Generate with your new generator
npm test          # Should pass with generated code
```

## Object Parameters Requirement (MANDATORY)

### **ALL Generated Functions MUST Use Object Parameters**

Generated callback functions and multi-parameter functions MUST use object parameters:

```typescript
// ✅ REQUIRED: Object parameter callbacks
const callbackType = `callback: (params: {error?: Error, data?: ${messageType}}) => void`;

// ✅ REQUIRED: Object parameter function calls
const functionCall = `callback({error: undefined, data: result});`;

// ✅ REQUIRED: Multi-parameter function signatures
const functionParameters = `({
  ${parameterList.join(', \n  ')}
}: {
  ${parameterTypeList.join(', \n  ')}
})`;
```

### **Generator Implementation Requirements**

- **Callback Signatures**: Always use `(params: {param1, param2, ...}) => void`
- **Function Calls**: Always use `callback({param1: value1, param2: value2})`
- **Multi-Parameter Functions**: Use object destructuring for 2+ parameters
- **Type Safety**: Provide explicit types for all object parameters

## Generator Pattern Compliance

### REQUIRED: Every generator must follow this exact pattern

```typescript
// 1. Define Zod schema with defaults
export const zodTypeScript[Name]Generator = z.object({
  id: z.string().optional().default('[name]-typescript'),
  preset: z.literal('[name]').default('[name]'),
  outputPath: z.string().optional().default('src/__gen__/[name]'),
  language: z.literal('typescript').optional().default('typescript'),
  // ... specific options with defaults
});

// 2. Create types
export type TypeScript[Name]Generator = z.input<typeof zodTypeScript[Name]Generator>;
export type TypeScript[Name]GeneratorInternal = z.infer<typeof zodTypeScript[Name]Generator>;

// 3. Implement core generator function
export async function generateTypescript[Name]Core(
  processedData: Processed[Name]Data,
  generator: TypeScript[Name]GeneratorInternal
): Promise<TypeScript[Name]RenderType> {
  const modelinaGenerator = new TypeScriptFileGenerator({
    ...defaultCodegenTypescriptModelinaOptions,
    // ... generator-specific options
  });
  // ... implementation
}

// 4. Implement main generator function
export async function generateTypescript[Name](
  context: TypeScript[Name]Context
): Promise<TypeScript[Name]RenderType> {
  // Process input based on type (asyncapi/openapi)
  // Call core generator
  // Return render type
}
```

## Zod Schema Patterns

### Generator Schema Requirements
- **Generator Schemas**: All generator schemas follow the pattern `zodTypeScript[GeneratorName]Generator`
- **Required Fields**: Always include `id`, `preset`, `outputPath`, `language` fields
- **Default Values**: Every optional field must have a sensible default using `.default()`
- **Discriminated Unions**: Use `preset` field as discriminator for generator unions

### Schema Registration
Add new schemas to appropriate unions in `src/codegen/types.ts`:
```typescript
export const zodAsyncAPITypeScriptGenerators = z.discriminatedUnion('preset', [
  zodTypeScriptPayloadGenerator,
  zodTypescriptParametersGenerator,
  // ... add new generators here
]);
```

### Type Safety Requirements
```typescript
// REQUIRED: Both input and internal types
export type TypeScript[Name]Generator = z.input<typeof zodTypeScript[Name]Generator>;
export type TypeScript[Name]GeneratorInternal = z.infer<typeof zodTypeScript[Name]Generator>;

// REQUIRED: Context interface
export interface TypeScript[Name]Context extends GenericCodegenContext {
  inputType: 'asyncapi' | 'openapi';
  asyncapiDocument?: AsyncAPIDocumentInterface;
  openapiDocument?: OpenAPIV3.Document | OpenAPIV2.Document | OpenAPIV3_1.Document;
  generator: TypeScript[Name]GeneratorInternal;
}

// REQUIRED: Render type interface
export type TypeScript[Name]RenderType = [Name]RenderType<TypeScript[Name]GeneratorInternal>;
```

## Modelina Integration Specifics

### File Generation Requirements
- **Always use `generateToFiles()`** method for better performance
- **Export Types**: Use `{exportType: 'named'}` for consistent exports
- **File Writing**: Generate directly to the file system, don't return strings

### Preset Configuration
```typescript
const modelinaGenerator = new TypeScriptFileGenerator({
  ...defaultCodegenTypescriptModelinaOptions,
  presets: [
    TS_DESCRIPTION_PRESET, // For documentation
    {
      preset: TS_COMMON_PRESET,
      options: {
        marshalling: true // Enable JSON marshalling
      }
    },
    {
      class: {
        additionalContent: ({content, model, renderer}) => {
          if (!generator.includeValidation) {
            return content;
          }
          return `${content}\n${generateTypescriptValidationCode({model, renderer})}`;
        }
      }
    }
  ],
  enumType: generator.enum || 'enum',
  mapType: generator.map || 'record',
  rawPropertyNames: generator.rawPropertyNames || false,
  useJavascriptReservedKeywords: generator.useForJavaScript !== false
});
```

### Validation Integration
- **Validation Code**: Add validation using `generateTypescriptValidationCode()` when `includeValidation` is true
- **Constraints**: Use `typeScriptDefaultPropertyKeyConstraints` for property name handling
- **Custom Presets**: Add generator-specific presets for custom behavior

## Generator Implementation Requirements

### Core Function Pattern
```typescript
export async function generateTypescript[Name]Core(
  processedData: Processed[Name]Data,
  generator: TypeScript[Name]GeneratorInternal
): Promise<TypeScript[Name]RenderType> {
  // 1. Create Modelina generator with proper configuration
  const modelinaGenerator = new TypeScriptFileGenerator({
    ...defaultCodegenTypescriptModelinaOptions,
    // Generator-specific options
  });

  // 2. Process each data category
  const results = {};
  
  // 3. Generate models using processedData
  for (const [key, schemaData] of Object.entries(processedData.items)) {
    if (schemaData) {
      const models = await modelinaGenerator.generateToFiles(
        schemaData.schema,
        generator.outputPath,
        {exportType: 'named'},
        true
      );
      results[key] = models[0]; // Use first model as primary
    }
  }

  // 4. Return structured result
  return {
    [resultKey]: results,
    generator
  };
}
```

### Main Function Pattern
```typescript
export async function generateTypescript[Name](
  context: TypeScript[Name]Context
): Promise<TypeScript[Name]RenderType> {
  const {asyncapiDocument, openapiDocument, inputType, generator} = context;

  let processedData: Processed[Name]Data;

  // REQUIRED: Input processing switch
  switch (inputType) {
    case 'asyncapi':
      if (!asyncapiDocument) {
        throw new Error('Expected AsyncAPI input, was not given');
      }
      processedData = await processAsyncAPI[Name](asyncapiDocument);
      break;
    case 'openapi':
      if (!openapiDocument) {
        throw new Error('Expected OpenAPI input, was not given');
      }
      processedData = processOpenAPI[Name](openapiDocument);
      break;
    default:
      throw new Error(`Unsupported input type: ${inputType}`);
  }

  // Call core generator
  return generateTypescript[Name]Core(processedData, generator);
}
```

## File System Conventions

### Output Path Patterns
- **Default Paths**: Use `src/__gen__/[generator-type]` pattern
- **File Extensions**: Generate `.ts` files for TypeScript
- **Index Files**: Create barrel exports in `index.ts` files
- **Directory Structure**: Maintain consistent directory structure in generated code

### File Organization
```
src/__gen__/
├── payloads/
│   ├── index.ts              # Barrel exports
│   ├── [ModelName].ts        # Individual model files
│   └── ...
├── parameters/
├── headers/
└── ...
```

## Error Handling Requirements

### Input Validation
```typescript
// REQUIRED: Document existence check
if (!asyncapiDocument) {
  throw new Error('Expected AsyncAPI input, was not given');
}

// REQUIRED: Schema validation
if (!schemaData.schema) {
  Logger.warn(`No schema found for ${itemName}, skipping generation`);
  continue;
}
```

### Context Information
- Include model names, property names, and file paths in error messages
- Use `Logger` from `LoggingInterface.ts` for consistent logging
- Provide warnings for ambiguous situations (e.g., multiple discriminator properties)

## Protocol Support (Channels Generator)

### Supported Protocols
- **NATS**: Message streaming and request/reply patterns
- **Kafka**: Event streaming with consumer groups
- **MQTT**: IoT messaging with QoS levels and user properties
- **AMQP**: Enterprise messaging with exchanges and queues
- **EventSource**: Server-sent events for real-time updates
- **HTTP Client**: RESTful API communication

### Function Types
- `publish`: Send messages to a channel/topic
- `subscribe`: Receive messages from a channel/topic
- `request`: Send request and wait for reply
- `reply`: Handle incoming requests and send responses

### Protocol-Specific Header Handling

Each protocol handles headers according to its specifications:

- **NATS**: Uses NATS headers (`msg.headers`) with `keys()` method or Object.entries fallback
- **Kafka**: Uses Kafka message headers (`message.headers`) as key-value pairs
- **AMQP**: Uses AMQP message properties (`msg.properties.headers`) 
- **MQTT**: Uses MQTT v5 user properties (`packet.properties.userProperties`)
- **EventSource**: Uses HTTP headers in the request

### MQTT-Specific Implementation Patterns

**MQTT v5 Requirements**:
- **CRITICAL**: MQTT v5 protocol version is required for user properties (headers) support
- MQTT v3.1.1 (default) does not support user properties
- Always configure clients with `{ protocolVersion: 5 }` for full functionality
- Both publish and subscribe operations require MQTT v5 for header transmission

**MQTT Subscribe Pattern**:
```typescript
// MQTT supports multiple event listeners without conflicts
mqtt.on('message', messageHandler);

// Extract headers from MQTT v5 user properties (requires protocolVersion: 5)
if (packet.properties && packet.properties.userProperties) {
  extractedHeaders = HeaderType.unmarshal(packet.properties.userProperties);
}

// Use object parameter callbacks (MANDATORY)
onDataCallback({err: undefined, msg: parsedMessage, parameters, headers: extractedHeaders, mqttMsg: packet});
```

**Header Processing**:
- **Publish**: Use `publishOptions.properties.userProperties` to send headers
- **Subscribe**: Use `packet.properties.userProperties` to receive headers  
- **Unmarshalling**: Always use `HeaderType.unmarshal()` not constructor for proper property mapping

**Channel Function Type Management**:
```typescript
// REQUIRED: Add new function types to enum
export enum ChannelFunctionTypes {
  MQTT_PUBLISH = 'mqtt_publish',
  MQTT_SUBSCRIBE = 'mqtt_subscribe', // ✅ Added for subscribe support
  // ... other types
}

// REQUIRED: Add to receiving function types array
const receivingFunctionTypes = [
  ChannelFunctionTypes.NATS_SUBSCRIBE,
  ChannelFunctionTypes.MQTT_SUBSCRIBE, // ✅ Required for proper operation mapping
  ChannelFunctionTypes.KAFKA_SUBSCRIBE,
  // ... other receiving types
];
```

### Code Generation Pattern
```typescript
// Generate protocol-specific functions with proper imports
const functionCode = `
export async function ${functionName}(${parameters}): Promise<${returnType}> {
  // Protocol-specific implementation
  ${protocolCode}
}
`;

// Add required dependencies
dependencies.push(...protocolDependencies);
```

## Specific Implementation Details

### String Manipulation
- Use utility functions from `src/codegen/generators/typescript/utils.ts`
- Use `castToTsType()` for JSON Schema to TypeScript type conversion
- Use `pascalCase()` for type names, maintain consistency
- Use `safeStringify()` for circular reference handling

### Configuration File Formats Support
- **ESM/MJS**: For JavaScript ES modules with `.mjs` extension
- **CJS**: For CommonJS modules with `.cjs` extension
- **TypeScript**: For `.ts` configuration files
- **JSON/YAML**: For static configurations without callbacks

## Forbidden Patterns in Generators
- ❌ **No `any` types** without explicit justification in comments
- ❌ **No `console.log`** - use `Logger` from `LoggingInterface.ts`
- ❌ **No hardcoded paths** - use configuration or constants
- ❌ **No synchronous file operations** - use async alternatives
- ❌ **No global variables** - pass state through function parameters
- ❌ **No direct string concatenation** for code generation - use template literals

## Performance Requirements
- **File Generation**: Always use `generateToFiles()` for better performance
- **Memory Management**: Avoid loading large schemas entirely into memory
- **Dependency Management**: Leverage Modelina's dependency management
- **Caching**: Cache processed schemas when processing multiple generators
