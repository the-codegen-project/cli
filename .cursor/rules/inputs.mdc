---
description: Input processing patterns and requirements for The Codegen Project CLI
globs: ["src/codegen/inputs/**/*.ts"]
alwaysApply: true
---

# Input Processing Rules

## Input Processing Pattern Compliance

### REQUIRED: Input processors must return standardized interfaces

```typescript
// Standard interface pattern for processed schema data
export interface Processed[Name]SchemaData {
  channelPayloads: Record<string, {schema: any; schemaId: string}>;
  operationPayloads: Record<string, {schema: any; schemaId: string}>;
  otherPayloads: {schema: any; schemaId: string}[];
}

// Alternative pattern for non-payload processors
export interface Processed[Name]Data {
  channel[Items]: Record<string, {schema: any; schemaId: string} | undefined>;
  operation[Items]: Record<string, {schema: any; schemaId: string} | undefined>;
  other[Items]: {schema: any; schemaId: string}[];
}
```

### Input Validation Pattern
```typescript
// REQUIRED: Input validation pattern
export function process[InputType][Name](document: [DocumentType]): Processed[Name]SchemaData {
  if (!document) {
    throw new Error('Expected [InputType] input, was not given');
  }
  
  // Processing logic here
  
  return {
    channelPayloads: {},
    operationPayloads: {},
    otherPayloads: []
  };
}
```

## Input Processing Patterns

### Separation of Concerns
- **Input Processors**: Located in `src/codegen/inputs/[input-type]/generators/`
- **Schema Extraction**: Extract schemas and return standardized `ProcessedXSchemaData` interfaces
- **Input Agnostic**: Core generators work with processed data, not raw input documents
- **Error Handling**: Validate inputs exist and throw descriptive errors

### File Organization
```
src/codegen/inputs/
├── asyncapi/
│   ├── parser.ts                    # Document parsing
│   ├── generators/
│   │   ├── payloads.ts             # Payload processing
│   │   ├── parameters.ts           # Parameter processing
│   │   ├── headers.ts              # Header processing
│   │   ├── types.ts                # Type processing
│   │   └── index.ts                # Barrel exports
│   └── index.ts                    # Main exports
├── openapi/
│   ├── parser.ts                   # Document parsing
│   ├── generators/
│   │   ├── payloads.ts             # Payload processing
│   │   ├── parameters.ts           # Parameter processing
│   │   ├── headers.ts              # Header processing
│   │   ├── types.ts                # Type processing
│   │   └── index.ts                # Barrel exports
│   └── index.ts                    # Main exports
```

## AsyncAPI Input Processing

### Document Parsing
```typescript
import {AsyncAPIDocumentInterface} from '@asyncapi/parser';

// REQUIRED: AsyncAPI document validation
export function processAsyncAPI[Name](
  asyncapiDocument: AsyncAPIDocumentInterface
): Processed[Name]SchemaData {
  if (!asyncapiDocument) {
    throw new Error('Expected AsyncAPI input, was not given');
  }

  // Extract channels, operations, and components
  const channels = asyncapiDocument.channels();
  const operations = asyncapiDocument.operations();
  const components = asyncapiDocument.components();

  // Process each category
  return {
    channelPayloads: processChannels(channels),
    operationPayloads: processOperations(operations),
    otherPayloads: processComponents(components)
  };
}
```

### Schema Extraction Patterns
```typescript
// Extract schema from AsyncAPI message
function extractMessageSchema(message: any): {schema: any; schemaId: string} | null {
  if (!message || !message.payload()) {
    return null;
  }

  const payload = message.payload();
  const schemaId = message.id() || generateSchemaId(message);
  
  return {
    schema: payload.json(),
    schemaId: pascalCase(schemaId)
  };
}

// Extract schema from AsyncAPI parameter
function extractParameterSchema(parameter: any): {schema: any; schemaId: string} | null {
  if (!parameter || !parameter.schema()) {
    return null;
  }

  return {
    schema: parameter.schema().json(),
    schemaId: pascalCase(parameter.id())
  };
}
```

## OpenAPI Input Processing

### Document Type Support
- **OpenAPI 2.0**: Support Swagger 2.0 specifications
- **OpenAPI 3.0**: Support OpenAPI 3.0 specifications  
- **OpenAPI 3.1**: Support OpenAPI 3.1 specifications with proper type guards

### Type Guards
```typescript
import {OpenAPIV2, OpenAPIV3, OpenAPIV3_1} from 'openapi-types';

// REQUIRED: OpenAPI document validation
export function processOpenAPI[Name](
  openapiDocument: OpenAPIV3.Document | OpenAPIV2.Document | OpenAPIV3_1.Document
): Processed[Name]SchemaData {
  if (!openapiDocument) {
    throw new Error('Expected OpenAPI input, was not given');
  }

  // Version-specific processing
  if (isOpenAPIV2(openapiDocument)) {
    return processOpenAPIV2[Name](openapiDocument);
  } else if (isOpenAPIV3(openapiDocument)) {
    return processOpenAPIV3[Name](openapiDocument);
  } else {
    return processOpenAPIV31[Name](openapiDocument);
  }
}
```

### Schema Extraction from Operations
```typescript
// Extract payload schemas from OpenAPI operations
function extractPayloadsFromOperations(
  paths: OpenAPIV3.PathsObject | OpenAPIV2.PathsObject | OpenAPIV3_1.PathsObject
): {
  requestPayloads: Record<string, {schema: any; schemaId: string}>;
  responsePayloads: Record<string, {schema: any; schemaId: string}>;
} {
  const requestPayloads: Record<string, {schema: any; schemaId: string}> = {};
  const responsePayloads: Record<string, {schema: any; schemaId: string}> = {};

  for (const [pathKey, pathItem] of Object.entries(paths)) {
    if (!pathItem) continue;

    for (const [method, operation] of Object.entries(pathItem)) {
      if (!operation || typeof operation !== 'object') continue;

      const operationId = operation.operationId ?? 
        `${method}${pathKey.replace(/[^a-zA-Z0-9]/g, '')}`;

      // Extract request schema
      const requestSchema = extractRequestSchema(operation);
      if (requestSchema) {
        requestPayloads[`${operationId}Request`] = requestSchema;
      }

      // Extract response schemas
      const responseSchemas = extractResponseSchemas(operation, operationId);
      Object.assign(responsePayloads, responseSchemas);
    }
  }

  return {requestPayloads, responsePayloads};
}
```

## Schema Processing Utilities

### Schema ID Generation
```typescript
// Generate consistent schema IDs
function generateSchemaId(source: any, fallback?: string): string {
  // Try to get ID from various sources
  const id = source.id?.() || 
             source.name?.() || 
             source.title ||
             source['x-modelina-name'] ||
             fallback ||
             'GeneratedSchema';
             
  return pascalCase(id);
}
```

### Schema Validation
```typescript
// Validate schema before processing
function validateSchema(schema: any, context: string): boolean {
  if (!schema) {
    Logger.warn(`No schema found for ${context}, skipping`);
    return false;
  }

  if (typeof schema !== 'object') {
    Logger.warn(`Invalid schema type for ${context}, expected object`);
    return false;
  }

  return true;
}
```

### Union Schema Creation
```typescript
// Create union schema from multiple schemas
function createUnionSchema(
  schemas: any[],
  baseId: string,
  hasStatusCodes: boolean = false
): any {
  if (schemas.length === 0) return null;
  if (schemas.length === 1) return schemas[0];

  const unionSchema = {
    $schema: 'http://json-schema.org/draft-07/schema',
    $id: baseId,
    title: baseId,
    oneOf: schemas,
    'x-modelina-has-status-codes': hasStatusCodes
  };

  return unionSchema;
}
```

## Error Handling Requirements

### Input Document Validation
```typescript
// REQUIRED: Always validate input documents exist
if (!document) {
  throw new Error('Expected [InputType] input, was not given');
}

// REQUIRED: Validate document structure
if (!document.channels && !document.paths) {
  throw new Error('Invalid document structure: missing channels or paths');
}
```

### Schema Processing Errors
```typescript
// Handle missing schemas gracefully
if (!schema) {
  Logger.warn(`No schema found for ${itemName}, skipping generation`);
  continue;
}

// Handle invalid schema types
if (typeof schema !== 'object') {
  Logger.error(`Invalid schema type for ${itemName}: ${typeof schema}`);
  throw new Error(`Cannot process non-object schema for ${itemName}`);
}
```

### Context Information in Errors
- Include document type (AsyncAPI/OpenAPI) in error messages
- Include channel/operation names when available
- Include schema paths for debugging
- Use descriptive error messages that help users fix issues

## Performance Considerations

### Memory Management
- Don't load entire documents into memory unnecessarily
- Process schemas incrementally when possible
- Use streaming for large document processing
- Cache parsed schemas when processing multiple generators

### Async Processing
```typescript
// REQUIRED: Use async processing for I/O operations
export async function processAsyncAPI[Name](
  asyncapiDocument: AsyncAPIDocumentInterface
): Promise<Processed[Name]SchemaData> {
  // Async processing logic
}
```

## Utility Functions

### Schema Transformation
```typescript
// Transform schema for Modelina compatibility
function transformSchemaForModelina(schema: any): any {
  // Add JSON Schema draft if missing
  if (!schema.$schema) {
    schema.$schema = 'http://json-schema.org/draft-07/schema';
  }

  // Ensure proper ID format
  if (schema.$id && !schema.$id.startsWith('#')) {
    schema.$id = `#/${schema.$id}`;
  }

  return schema;
}
```

### Component Processing
```typescript
// Extract schemas from document components
function extractComponentSchemas(
  components: any
): {schema: any; schemaId: string}[] {
  const componentSchemas: {schema: any; schemaId: string}[] = [];

  if (components?.schemas) {
    for (const [name, schema] of Object.entries(components.schemas)) {
      if (validateSchema(schema, `component.${name}`)) {
        componentSchemas.push({
          schema: transformSchemaForModelina(schema),
          schemaId: pascalCase(name)
        });
      }
    }
  }

  return componentSchemas;
}
```

## Required Imports
```typescript
// REQUIRED: Input processing imports
import {AsyncAPIDocumentInterface} from '@asyncapi/parser';
import {OpenAPIV2, OpenAPIV3, OpenAPIV3_1} from 'openapi-types';
import {Logger} from '../../../LoggingInterface';
import {pascalCase} from '../../generators/typescript/utils';
import {onlyUnique} from '../../utils';
```

## Forbidden Patterns in Input Processing
- ❌ **No direct document manipulation** - work with copies
- ❌ **No synchronous file operations** - use async alternatives
- ❌ **No hardcoded schema paths** - use dynamic extraction
- ❌ **No global state** - pass context through parameters
- ❌ **No `any` types** without proper validation
