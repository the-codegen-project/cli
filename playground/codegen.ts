import { TheCodegenConfiguration } from "@the-codegen-project/cli";
const config: TheCodegenConfiguration = {
  inputType: "asyncapi",
  inputPath: "./schema.json",
  language: "typescript",
  generators: [
    {
      preset: "channels",
      outputPath: "./__gen__/channels",
      protocols: ["nats"],
    },
    {
      preset: "payloads",
      outputPath: "./__gen__/payloads",
      serializationType: "json",
      model: "class",
      enum: "enum",
      map: "record",
      moduleSystem: "esm",
      useForJavaScript: true,
      rawPropertyNames: false,
    },
    {
      preset: "parameters",
      outputPath: "./__gen__/parameters",
      serializationType: "json",
    },
  ],
};
export default config;
