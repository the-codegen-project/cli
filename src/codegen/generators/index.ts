import {
  TypescriptParametersGenerator,
  generateTypescriptParameters,
  defaultTypeScriptParametersOptions,
  TypeScriptPayloadGenerator,
  generateTypescriptPayload,
  defaultTypeScriptPayloadGenerator,
  TypeScriptChannelsGenerator,
  generateTypeScriptChannels,
  defaultTypeScriptChannelsGenerator,
  TypeScriptClientGenerator,
  defaultTypeScriptClientGenerator,
  generateTypeScriptClient,
  TypeScriptChannelsContext,
  TypeScriptClientContext,
  TypeScriptPayloadContext,
  TypescriptParametersContext,
  TypeScriptPayloadGeneratorInternal,
  TypeScriptChannelsGeneratorInternal,
  TypeScriptClientGeneratorInternal,
  TypescriptParametersGeneratorInternal,
  TypeScriptparameterRenderType
} from './typescript';
import {defaultCustomGenerator, CustomGenerator} from './generic/custom';
import {realizedConfiguration} from '../configurations';
import {runGenerators} from '..';

export {
  TypeScriptChannelsGenerator,
  generateTypeScriptChannels,
  defaultTypeScriptChannelsGenerator,
  TypeScriptPayloadGenerator,
  generateTypescriptPayload,
  defaultTypeScriptPayloadGenerator,
  TypescriptParametersGenerator,
  generateTypescriptParameters,
  defaultTypeScriptParametersOptions,
  TypeScriptClientGenerator,
  defaultTypeScriptClientGenerator,
  generateTypeScriptClient,
  CustomGenerator,
  defaultCustomGenerator,
  TypeScriptChannelsContext,
  TypeScriptClientContext,
  TypeScriptPayloadContext,
  TypescriptParametersContext,
  TypeScriptPayloadGeneratorInternal,
  TypeScriptChannelsGeneratorInternal,
  TypeScriptClientGeneratorInternal,
  TypescriptParametersGeneratorInternal,
  TypeScriptparameterRenderType
};

/**
 * Load the configuration and run the generator
 *
 * @param configFile
 */
export async function generateWithConfig(configFile: string | undefined) {
  const context = await realizedConfiguration(configFile);
  await runGenerators(context);
}
