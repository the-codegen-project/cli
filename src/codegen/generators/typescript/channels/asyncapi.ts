import {ChannelFunctionTypes} from './types';

type Action = 'send' | 'receive' | 'subscribe' | 'publish';
const sendingFunctionTypes = [
  ChannelFunctionTypes.NATS_JETSTREAM_PUBLISH,
  ChannelFunctionTypes.NATS_PUBLISH,
  ChannelFunctionTypes.NATS_REQUEST,
  ChannelFunctionTypes.MQTT_PUBLISH,
  ChannelFunctionTypes.KAFKA_PUBLISH,
  ChannelFunctionTypes.AMQP_EXCHANGE_PUBLISH,
  ChannelFunctionTypes.AMQP_QUEUE_PUBLISH,
  ChannelFunctionTypes.EVENT_SOURCE_EXPRESS
];
const receivingFunctionTypes = [
  ChannelFunctionTypes.NATS_JETSTREAM_PULL_SUBSCRIBE,
  ChannelFunctionTypes.NATS_JETSTREAM_PUSH_SUBSCRIBE,
  ChannelFunctionTypes.NATS_REPLY,
  ChannelFunctionTypes.NATS_SUBSCRIBE,
  ChannelFunctionTypes.KAFKA_SUBSCRIBE,
  ChannelFunctionTypes.EVENT_SOURCE_FETCH
];

// eslint-disable-next-line sonarjs/cognitive-complexity
export function shouldRenderFunctionType(
  givenFunctionTypes: ChannelFunctionTypes[] | undefined,
  functionTypesToCheckFor: ChannelFunctionTypes | ChannelFunctionTypes[],
  action: Action,
  reverseOperation: boolean
) {
  const listToCheck = [
    ...(Array.isArray(functionTypesToCheckFor)
      ? functionTypesToCheckFor
      : [functionTypesToCheckFor])
  ];
  const hasSendingOperation = action === 'send' || action === 'subscribe';
  const hasReceivingOperation = action === 'receive' || action === 'publish';
  const hasFunctionMappingConfig = givenFunctionTypes !== undefined;
  const checkForSending = listToCheck.some((item) =>
    sendingFunctionTypes.includes(item)
  );
  const checkForReceiving = listToCheck.some((item) =>
    receivingFunctionTypes.includes(item)
  );
  const hasFunctionType = (givenFunctionTypes ?? []).some((item) =>
    listToCheck.includes(item)
  );
  if (hasFunctionMappingConfig) {
    if (hasFunctionType) {
      const renderForSending = checkForSending && hasSendingOperation;
      const renderForReceiving = checkForReceiving && hasReceivingOperation;
      return renderForSending || renderForReceiving;
    }
    return false;
  }

  if (reverseOperation) {
    const renderForSending = hasSendingOperation && checkForReceiving;
    const renderForReceiving = hasReceivingOperation && checkForSending;
    return renderForSending || renderForReceiving;
  }

  const renderForSending = checkForSending && hasSendingOperation;
  const renderForReceiving = checkForReceiving && hasReceivingOperation;
  return renderForSending || renderForReceiving;
}
