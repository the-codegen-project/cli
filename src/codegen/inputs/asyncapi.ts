import AsyncapiParser from "@asyncapi/parser";

import { AvroSchemaParser } from '@asyncapi/avro-schema-parser';
import { OpenAPISchemaParser } from '@asyncapi/openapi-schema-parser';
import { RamlDTSchemaParser } from '@asyncapi/raml-dt-schema-parser';
import { ProtoBuffSchemaParser } from '@asyncapi/protobuf-schema-parser';
import { RunGeneratorContext } from "../types.js";

const parser = new AsyncapiParser.Parser({
	ruleset: {
		core: false,
		recommended: false
	},
	schemaParsers: [
		AvroSchemaParser(),
		OpenAPISchemaParser(),
		RamlDTSchemaParser(),
		ProtoBuffSchemaParser(),
	]
});

export async function loadAsyncapi(context: RunGeneratorContext) {
	const document = await AsyncapiParser.fromFile(parser, context.documentPath).parse()
	if(document.diagnostics.length > 0) {
        throw new Error(`Could not load AsyncAPI document, errors was: ${JSON.stringify(document.diagnostics)}`);
	}
	return document.document
}