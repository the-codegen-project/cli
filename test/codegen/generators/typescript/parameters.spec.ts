import path from "node:path";
import { generateTypescriptParameters } from "../../../../src/codegen/generators";
import { loadAsyncapiDocument } from "../../../../src/codegen/inputs/asyncapi";

describe('payloads', () => {
  describe('typescript', () => {
    it('should work with AsyncAPI that contains parameters', async () => {
      const parsedAsyncAPIDocument = await loadAsyncapiDocument(path.resolve(__dirname, '../../../configs/parameters.yaml'));
      
      const renderedContent = await generateTypescriptParameters({
        generator: {
          serializationType: 'json',
          outputPath: path.resolve(__dirname, './output'),
          preset: 'parameters',
          language: 'typescript',
          dependencies: [],
          id: 'test'
        },
        inputType: 'asyncapi',
        asyncapiDocument: parsedAsyncAPIDocument,
        dependencyOutputs: { }
      });
      expect(renderedContent.channelModels['single_parameter']?.result).toMatchSnapshot();
      expect(renderedContent.channelModels['second_parameter']?.result).toMatchSnapshot();
    });
    it('should work with no channels', async () => {
      const parsedAsyncAPIDocument = await loadAsyncapiDocument(path.resolve(__dirname, '../../../configs/parameters-no-channels.yaml'));
      
      const renderedContent = await generateTypescriptParameters({
        generator: {
          serializationType: 'json',
          outputPath: path.resolve(__dirname, './output'),
          preset: 'parameters',
          language: 'typescript',
          dependencies: [],
          id: 'test'
        },
        inputType: 'asyncapi',
        asyncapiDocument: parsedAsyncAPIDocument,
        dependencyOutputs: { }
      });
      expect(Object.keys(renderedContent.channelModels).length).toEqual(0);
    });
  });
});
