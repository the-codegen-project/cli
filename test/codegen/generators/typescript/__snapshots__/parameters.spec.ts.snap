// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payloads typescript should work with AsyncAPI that contains parameters 1`] = `
"
class SingleParameterParameters {
  private _firstParameter: string;

  constructor(input: {
    firstParameter: string,
  }) {
    this._firstParameter = input.firstParameter;
  }

  get firstParameter(): string { return this._firstParameter; }
  set firstParameter(firstParameter: string) { this._firstParameter = firstParameter; }


  /**
   * Realize the channel/topic with the parameters added to this class.
   */
  public getChannelWithParameters(channel: string) {
    channel = channel.replace(/\\{first_parameter\\}/g, this.firstParameter);
    return channel;
  }
  
  public static createFromChannel(msgSubject: string, channel: string, regex: RegExp): SingleParameterParameters {
    const parameters = new SingleParameterParameters({firstParameter: ''});
  const match = msgSubject.match(regex);
  const sequentialParameters: string[] = channel.match(/\\{(\\w+)\\}/g) || [];

  if (match) {
    const firstParameterMatch = match[sequentialParameters.indexOf('{first_parameter}')+1];
        if(firstParameterMatch && firstParameterMatch !== '') {
          parameters.firstParameter = firstParameterMatch as any
        } else {
          throw new Error(\`Parameter: 'firstParameter' is not valid. Abort! \`) 
        }
  } else {
    throw new Error(\`Unable to find parameters in channel/topic, topic was \${channel}\`)
  }
  return parameters;
  }
}
export { SingleParameterParameters };"
`;

exports[`payloads typescript should work with AsyncAPI that contains parameters 2`] = `undefined`;
